### ISSUE TITLE
Customer is unable to login with microsoft authentication provide from vscode.dev due to Conditional Access Policies

### ISSUE THREAD

**Issue by sevillal** [microsoft/vscode#197389]
Original: https://github.com/microsoft/vscode-internalbacklog/issues/4478

## From Azure ML extension team (Sergio)

> A customer using the Azure ML remote connection in web reported that it is unable to login and that is stuck in an authentication loop, so after the customer enters its user and password the redirection to vscode.dev happens but eventually is taken back to the microsoft authentication prompt.
> 
> These are my findings on the initial triage:
> 
> The Azure ML web opener calls the microsoft authentication provider getSessions
> Since there are not valid sessions found, the login process begins, and customer is taken to the microsoft login prompt
> User enter the user password information and redirection starts
> User gets 400 when the anything route calls the https://vscode.dev/codeExchangeProxyEndpoints/microsoft/{TENANT}/oauth2/v2.0/token API as part of the tryCompleteAuthExchage
> The Azure ML web opener calls the getSessions again and the cycle repeats
> Follow ups:
> 
> Why the 400 happens in the first place?
> How can we get more information about the cause of the 400 when exchanging the code?
> Can we make it available in telemetry? Can the webOpener get this information to log telemetry or display error information?
> Next actions:
> 
> Azure ML extension to implement something to stop the loop on the web opener.

## From VS Code team (Tyler)

> The code for this is pretty simple:
> 
> the code for the token proxy route just returns the results of the call: https://github.com/microsoft/vscode-dev/blob/main/server/codeExchangeProxies.ts#L48-L53
> the code that calls the endpoint console.error's the result of the call: https://github.com/microsoft/vscode-dev/blob/main/client/contrib/microsoftAuth.ts#L449-L484 and https://github.com/microsoft/vscode-dev/blob/main/client/contrib/microsoftAuth.ts#L164
> So my guess is either:
> 
> Identity outage?
> Maybe the thing causing the Bad Request doesn't ever make it to the codeExchangeProxy endpoint... @joaomoreno do you know of any reason this might happen?
> 

## From Azure ML extension team (Sergio)

> This is the response we are getting for request https://vscode.dev/codeExchangeProxyEndpoints/microsoft/{TENANT}/oauth2/v2.0/token:
> 
> {"error":"invalid_grant","error_description":"AADSTS53003: Access has been blocked by Conditional Access policies. The access policy does not allow token issuance.\r\nTrace ID: 2f163c61-770b-4d3f-87de-df6430161700\r\nCorrelation ID: 3e5968cc-344d-49c1-b9ae-961ec2a34eab\r\nTimestamp: 2023-07-28 15:53:36Z","error_codes":[53003],"timestamp":"2023-07-28 15:53:36Z","trace_id":"2f163c61-770b-4d3f-87de-df6430161700","correlation_id":"3e5968cc-344d-49c1-b9ae-961ec2a34eab","error_uri":"https://login.microsoftonline.com/error?code=53003","suberror":"message_only"}
> Conditinal Access policies is a known limitation in vscode.dev because of the way it acquires tokens using the codeExchangeProxyEndpoint route. This is done to work around CORS constraints.
> According to VS code some users have reported using Edge as their browser got around the issue
> This is a problem specific to vscode.dev and it should work fine in desktop because there are not CORS issues.
> 
> Next steps:
> 
> - Ask customers to try Edge as their browser
> - Ask customers to disable the Conditional Access blocking the authorization
> - Measure how many customers are impacted and determine if this should be prioritized in web

**Comment by tyaginidhi** [comment]
Around authentication API error response handling - we reached out to Conditional Access team and here is a recommendation from them:

Nidhi, Tyler, our recommendation is:
Transition your authentication logic to use MSAL lib. This library should handle all edge cases for client errors and flows.
Handle MSAL errors properly and propagate them to the user so they could remediate if something happens.
for instance, if a user need to complete MFA, you should be able to redirect him to the auth endpoint to complete it. MSAL lib will assist your client to do that.

You can follow this guide:
[Handle errors and exceptions in MSAL.NET - Microsoft Authentication Library for .NET | Microsoft Learn](https://learn.microsoft.com/en-us/entra/msal/dotnet/advanced/exceptions/msal-error-handling)

@TylerLeonhardt  we are currently not using MSAL, correct? Can you confirm?

**Comment by tyaginidhi** [comment]
Looking at the [another issue associated above](https://github.com/microsoft/vscode-dev/issues/1036) - seems we are already planning to move to MSAL - that should also solve (to an extent) - the exception surfacing to client for us.

**Comment by TylerLeonhardt** [comment]
This is progressing. In theory, all the work in Microsoft/vscode is done.

Here's what's left:
* vscode.dev -> adopt MSAL (currently blocked on https://github.com/AzureAD/microsoft-authentication-library-for-js/issues/6269 I am working with the team to figure out a fix... hopefully it'll be resolved next week)
* We need to add a couple of redirect URIs to VS Code's Client Id: (`https://vscode.dev/hi` and `https://insiders.vscode.dev/hi`)... this is taking longer than expected because the new process for managing first party apps is really _really_ rough. There's also an AzDO outage at the moment that prevented me from getting as far as I wanted today.

Hopefully, this will all be addressed next week.

**Comment by TylerLeonhardt** [comment]
Ok so I've worked around the issue... now I'm on to another blocker. MSAL.js's popup flow doesn't work if the page has Cross-Origin-Opener-Policy set to `same-origin`... which we do for certain features to work (SharedArrayBuffers).

I've been working tirelessly on a workaround for this... the best I could come up with so far is owning my own popup then using MSAL.js's Redirect flow inside of it... but I've hit many many paper cuts along the way that have made it challenging.

I'm hoping to get some assistance from MSAL folks soon to work through these challenges.

**Comment by TylerLeonhardt** [comment]
Alright so we've got a workaround for the `same-origin` stuff... here's the issue on MSAL.js that talks about it.
https://github.com/AzureAD/microsoft-authentication-library-for-js/issues/6829

but the journey is not over... now I am dealing with this adoption blocker: https://github.com/AzureAD/microsoft-authentication-library-for-js/issues/6830

**Comment by TylerLeonhardt** [comment]
That said, I did push a small fix to our existing code that may solve this issue... it's in insiders.vscode.dev

**Comment by TylerLeonhardt** [comment]
Although the issue I linked is closed, I am following up with MSAL internally to come up with a plan. Sadly, don't expect much here soon as I need a solution provided by MSAL.

**Comment by tbombach** [comment]
Summarizing the offline meeting about this, we discussed 2 possible solutions for introducing MSAL.js into the Azure ML scenario to allow auth to succeed for customers with Conditional Access Policies:

**Solution 1**: Enable MSAL.js for all users for the Azure ML route or with a special query parameter.

**Solution 2**: MSAL.js as a fallback for users for whom the current hybrid vscode.dev auth flow fails.

We understand that our scenario needs to prepare for the eventuality that Solution 1 will likely be the only auth flow available (particularly the experience of interactive re-auth when tokens expire). Our focus here is rolling out the MSAL.js auth flow with minimal disruption to current users and to help with testing this codepath for it to be applied more broadly.

**Ask:**

Our preference would be Solution 2, as it would help accelerate the transition to MSAL.js since this would test the code with any customer running into the Conditional Access issue. Solution 1 would require us to enable MSAL.js for a very small initial set of users and we would roll this out slower since it would impact current users who are not experiencing any issues.

Questions for @TylerLeonhardt and @sbatten:
1. Would Solution 2 be an acceptable step towards the eventual solution of MSAL.js as the only auth flow?
2. Is there a difference in timeline between the two possible solutions?

Thank you again for all the investigation and time dedicated to moving this forward!

**Comment by TylerLeonhardt** [comment]
@tbombach and team, and I, have talked offline and route-specific MSAL enablement is how we wanna proceed for now. That is now live (in insiders.vscode.dev and next week vscode.dev) and able to for you to take advantage of in your route.

Please give it a try and let me know how it goes.

**Comment by connor4312** [comment]
@TylerLeonhardt are there steps for this?

**Comment by TylerLeonhardt** [comment]
I'll drive the verification for this, you can ignore it. I'll leave steps-needed so folks don't open it.

**Comment by TylerLeonhardt** [comment]
Marking as verified because Power Portals team says the flow works. We'll move forward and hopefully customers hit by CA policies are also helped.
